File: discover-client.dart
// @dart=2.9
import './discover_client_flutter/lib/client.dart';
import 'dart:io';

void main() {
  Client client = new Client();
  var query = "urn:schemas-sbtvd-org:service:GingaCCWebServices:1";
  var found = false;
  client.search(query, (data) async {
    if (found) return;
    found = true;
    print("-- m-search response headers ... " + data);
    String locationUrl = new RegExp(r"LOCATION: *(.*)\s", caseSensitive: false)
        .firstMatch(data)[1];
    HttpClient client = new HttpClient();
    client.getUrl(Uri.parse(locationUrl)).then((HttpClientRequest request) {
      return request.close();
    }).then((HttpClientResponse response) {
      print("-- response from " + locationUrl);
      print("-- GingaCC-Server-BaseURL = " +
          response.headers["GingaCC-Server-BaseURL"][0]);
      print("-- GingaCC-Server-SecureBaseURL = " +
          response.headers["GingaCC-Server-SecureBaseURL"][0]);
    });
  });
}

File: discover-client.js
// log funcs
function logStatusObject(name, obj) {
  if (obj === null) {
    console.log('-- ' + name);
    return;
  }
  console.log('-- ' + name + ' = ' + JSON.stringify(obj, null, 2));
}
function logStatus(msg) {
  console.log('-- ' + msg);
}

// ssdp m-search
const assert = require('assert').strict;
const url = require('url');
const http = require('http');
var SSDPClient = require('node-ssdp').Client
const GINGA_SSDP_TARGET = 'urn:schemas-sbtvd-org:service:GingaCCWebServices:1'
const BaseURL = 'GingaCC-Server-BaseURL'
const SecureBaseURL = 'GingaCC-Server-SecureBaseURL'

client = new SSDPClient({
  reuseAddr: false,
  explicitSocketBind: true // important: force use all interfaces
});

client.on('advertise-alive', function (headers) {
  console.log('advertise-alive');
  logStatusObject('advertise-alive headers ', headers);
});

client.on('advertise-bye', function (headers) {
  console.log('advertise-bye');
  logStatusObject('advertise-bye headers ', headers);
});

client.on('response', function (headers, statusCode, rinfo) {
  console.log('response');
  logStatus('m-search response from ' + rinfo.address);
  logStatusObject('response headers ', headers);
  assert(statusCode == 200);

  // get LOCATION header
  var location = 'undefined'
  if (headers['LOCATION'])
    location = headers['LOCATION'];

  // perform request to LOCATION
  const uri = new url.URL(location);
  const options = {
    hostname: uri.hostname,
    port: uri.port,
    path: uri.pathname,
    method: 'GET'
  }
  var req = http.request(options, response => {
    var headers = response.headers;
    logStatus('get response from ' + location);
    logStatusObject('headers ', headers);
    logStatus('GingaCC-Server BaseURL=' + headers[BaseURL.toLowerCase()])
    logStatus('GingaCC-Server BaseURL=' + headers[SecureBaseURL.toLowerCase()])
    console.log()
  });
  req.end()
});

var search = function () {
  logStatus(`m-search for ${GINGA_SSDP_TARGET}`);
  client.stop()
  client.search(GINGA_SSDP_TARGET);
  // client.search('ssdp:all')
}

logStatus(`m-search for ${GINGA_SSDP_TARGET}`);
search();
setInterval(search, 4000);
